File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    let answer = [];
    
    const total = {};
    for (let i = 0; i < genres.length; i++) {
        total[genres[i]] = (total[genres[i]] || 0) + plays[i];
    }
    
    const sortedGenres = Object.keys(total).sort((a, b) => total[b] - total[a]);
    
    const playCount = {};
    for (let i = 0; i < genres.length; i++) {
        playCount[genres[i]] = (playCount[genres[i]] || []).concat(i);
    }
    
    for (let genre in playCount) {
        playCount[genre].sort((a, b) => plays[b] - plays[a]);
        playCount[genre] = playCount[genre].slice(0, 2);
    }
    
    for (const genre of sortedGenres) {
        answer = [...answer, ...playCount[genre]];
    }
    
    
    return answer;
}

File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    dict = {}
    count_dict={}

    for i, genre in enumerate(genres):
        if genre not in dict:
            dict[genre] = list()
            count_dict[genre] = 0

        dict[genre].append((plays[i], i))
        count_dict[genre] += plays[i]

    sorted_genre = sorted(count_dict.items(), key = lambda x:x[1], reverse = True)
 
    for genre, _ in sorted_genre:
        dict[genre].sort(key = lambda x: (-x[0], x[1]))
        answer.append(dict[genre][0][1])
        if (len(dict[genre]) > 1):
            answer.append(dict[genre][1][1])

    return answer
